File: Array/0238-ProductOfArray/Solution.py
from typing import List

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [1] * n
        right_prod = 1
        # Calculate the product of all elements to the left of each element
        for i in range(1, n):
            result[i] = result[i-1] * nums[i-1]
        # Calculate the product of all elements to the right of each element and multiply with the left product
        for i in range(n-1, -1, -1):
            result[i] *= right_prod
            right_prod *= nums[i]
        return result

"""
Space Complexity: O(1)
Time Complexity: O(n)
"""

File: Array/0238-ProductOfArray/Solution.java
import java.util.Arrays;

class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        Arrays.fill(result, 1);
        int rightProd = 1;
        // Calculate the product of all elements to the left of each element
        for (int i = 1; i < n; i++) {
            result[i] = result[i - 1] * nums[i - 1];
        }
        // Calculate the product of all elements to the right of each element and multiply with the left product
        for (int i = n - 1; i >= 0; i--) {
            result[i] *= rightProd;
            rightProd *= nums[i];
        }
        return result;
    }
}
